pkg/controllers/machinesetmigration/machineset_migration_controller.go

```diff
@@ -0,0 +1,391 @@
+		// ClusterAPI is currently authoritative, check its generation against the synchronized one.
+		// TODO: InfraMachineTemplate.
+		capiMachineSet := &capiv1beta1.MachineSet{}
+		if err := r.Get(ctx, client.ObjectKey{Namespace: consts.DefaultManagedNamespace, Name: mapiMachineSet.Name}, capiMachineSet); err != nil {
+			return false, fmt.Errorf("failed to get CAPI machine set: %w", err)
+		}
+
+		return mapiMachineSet.Status.SynchronizedGeneration == capiMachineSet.Generation, nil
+	default:
+		// Any other value is disallowed by the openAPI schema validation.
+	}
+
+	return false, nil
+}
+
+// applyStatusAuthoritativeAPIWithPatch updates the authoritativeAPI field
+// using a server side apply patch. We do this to force ownership of the field.
+func (r *MachineSetMigrationReconciler) applyStatusAuthoritativeAPIWithPatch(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, authority machinev1beta1.MachineAuthority) error {
+	statusAc := machinev1applyconfigs.MachineSetStatus().WithAuthoritativeAPI(authority)
+	msAc := machinev1applyconfigs.MachineSet(mapiMachineSet.GetName(), mapiMachineSet.GetNamespace()).
+		WithStatus(statusAc)
+
+	if err := r.Status().Patch(ctx, mapiMachineSet, util.ApplyConfigPatch(msAc), client.ForceOwnership, client.FieldOwner(controllerName+"-AuthoritativeAPI")); err != nil {
+		return fmt.Errorf("failed to patch MAPI machine set status with authoritativeAPI: %w", err)
+	}
+
+	return nil
+}
+
+// ensureFinalizerOnNewAuthoritativeAPI adds a finalizer to the resource if required.
+// If the finalizer already exists, this function should be a no-op.
+// If the finalizer is added, the function will return true so that the reconciler can requeue the object.
+// Adding the finalizer in a separate reconcile ensures that spec updates are separate from status updates.
+func (r *MachineSetMigrationReconciler) ensureFinalizerOnNewAuthoritativeAPI(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet) (bool, error) {
+	var newAuthoritativeResource client.Object
+	switch mapiMachineSet.Spec.AuthoritativeAPI {
+	case machinev1beta1.MachineAuthorityClusterAPI:
+		// TODO: InfraMachineTemplate.
+		capiMachineSet := &capiv1beta1.MachineSet{}
+		if err := r.Get(ctx, client.ObjectKey{Namespace: consts.DefaultManagedNamespace, Name: mapiMachineSet.Name}, capiMachineSet); err != nil {
+			return false, fmt.Errorf("failed to get CAPI machine set: %w", err)
+		}
+
+		newAuthoritativeResource = capiMachineSet
+	case machinev1beta1.MachineAuthorityMachineAPI:
+		newAuthoritativeResource = mapiMachineSet
+	default:
+		// Any other value is disallowed by the openAPI schema validation.
+	}
+
+	// Check if we need to add the finalizer.
+	for _, finalizer := range newAuthoritativeResource.GetFinalizers() {
+		if finalizer == migrationControllerFinalizer {
+			return false, nil
+		}
+	}
+
+	newAuthoritativeResource.SetFinalizers(append(newAuthoritativeResource.GetFinalizers(), migrationControllerFinalizer))
```

Comment:
I think in terms of finalizers, the migration controller is supposed to ensure that if the MAPI resource doesn't have the MAPI finalizer, it adds it, and if the CAPI resource doesn't have the expected CAPI finalizer, it adds that.

I think the sync controller has its own finalizer that it is supposed to add to resources to enable deletion synchronisation, but I don't think a migration specific finalizer is needed